### PR TITLE
Add Vector4 types

### LINKED ISSUE #143: Remove `AxisDefs` enum
The `enum AxisDefs` was initially intended to provide symbolic access to space point information. It is conflicting with multiple other ways to this information and should therefore be removed to avoid confusion.

* Full space points should use the `SpacePointVector` definition and the related enum. The internal layout can be set by the user (e.g. t,x,y,z instead of x,y,z,t) and could thus be inconsistent with `AxisDefs`.
* `Eigen::Vector`s always provide `.x()`, `.y()`, `.z()`, `.w()` accessors for all appropriately-sized types. The index positions are hardcoded. This is the most often used way to access components (by far).

To avoid any confusion and have always a single well-defined way of accessing information. true space-points must use `SpacePointVector` and related enums. Lower-dimensional vectors must use the `.{x,y,z}()` accessors. `enum AxisDefs` should be removed.

### PR DESCRIPTION
**Pull request by msmk0**
Add `Vector4{F,D}` types and related transformations and extend the `CoordinateIndices` enum. A general guideline is added to the documentation outlining the preferred way to consistently access vector/matrix components.

This prepares for the removal of the dedicated `SpacePoint` type and closes #143.

### CODE REVIEW

File: Core/include/Acts/Utilities/Definitions.hpp [hunk 1]
```diff
@@ -111,36 +111,88 @@ using ActsRowVectorX = Eigen::Matrix<T, 1, Eigen::Dynamic>;
...
 using ActsRowVectorXf = ActsRowVectorX<float>;
 
-using Rotation3D = Eigen::Quaternion<double>;
-using Translation3D = Eigen::Translation<double, 3>;
-using AngleAxis3D = Eigen::AngleAxisd;
-using Transform3D = Eigen::Affine3d;
-using Vector3D = Eigen::Matrix<double, 3, 1>;
-using Vector2D = Eigen::Matrix<double, 2, 1>;
-using RotationMatrix3D = Eigen::Matrix<double, 3, 3>;
+/// @defgroup coordinates Fixed-size vector/matrix types for coordinates
```
Comment (FabianKlimpel) [inline Core/include/Acts/Utilities/Definitions.hpp:114]: This is a too specific description. Could you rephrase it such that these represent `e.g.` coordinates etc.?

File: Core/include/Acts/Utilities/Definitions.hpp [hunk 2]
```diff
@@ -111,36 +111,88 @@ using ActsRowVectorX = Eigen::Matrix<T, 1, Eigen::Dynamic>;
...
+using RotationMatrix4D = Eigen::Matrix<double, 4, 4>;
+// pure rotation transformations. only available in 2d and 3d
+using Rotation2F = Eigen::Rotation2D<float>;
+using Rotation3F = Eigen::Quaternion<float>;
+using AngleAxis3F = Eigen::AngleAxis<float>;
+using Rotation2D = Eigen::Rotation2D<double>;
+using Rotation3D = Eigen::Quaternion<double>;
+using AngleAxis3D = Eigen::AngleAxis<double>;
+// combined affine transformations. types are chosen for better data alignment:
+// - 2d affine stored as 2x3 matrix
```
Comment (FabianKlimpel) [inline Core/include/Acts/Utilities/Definitions.hpp:186]:
```suggestion
// - 2d affine compact stored as 2x3 matrix
```

File: Core/include/Acts/Utilities/Definitions.hpp (hunk 1 above)
Comment (msmk0) [reply Core/include/Acts/Utilities/Definitions.hpp]: Fixed.

File: Core/include/Acts/Utilities/Definitions.hpp (hunk 2 above)
Comment (msmk0) [reply Core/include/Acts/Utilities/Definitions.hpp]: Fixed.

**Comment by msmk0** [comment]
> Will there be a follow up PR that changes the accessors in the existing code?

There will be one at some point, but this is low on my priority list. This might a good *first issue* for newcomers.

**Comment by msmk0** [comment]
Switched to draft. I realize I need to add some more cleanup.

**Comment by msmk0** [comment]
@FabianKlimpel I realized that this only makes sense if the updated types and enum are consistently used throughout the code base. I added the obvious changes to the PR. Could you have another look?

File: Fatras/include/ActsFatras/EventData/Particle.hpp [hunk 3]
```diff
@@ -132,30 +130,26 @@ class Particle {
...
-  ///
-  /// The component order is [px,py,pz,E].
   Vector4 momentum4() const {
     Vector4 mom4;
     // stored direction is always normalized
-    mom4[0] = m_absMomentum * m_unitDirection[0];
-    mom4[1] = m_absMomentum * m_unitDirection[1];
-    mom4[2] = m_absMomentum * m_unitDirection[2];
-    mom4[3] = energy();
+    mom4[Acts::eMom0] = m_absMomentum * m_unitDirection[0];
```
Comment (FabianKlimpel) [inline Fatras/include/ActsFatras/EventData/Particle.hpp:142]:
```suggestion
    mom4[Acts::eMom0] = m_absMomentum * m_unitDirection[Acts::ePos0];
```

File: Fatras/include/ActsFatras/EventData/Particle.hpp [hunk 4]
```diff
@@ -132,30 +130,26 @@ class Particle {
...
+    mom4[Acts::eMom1] = m_absMomentum * m_unitDirection[1];
```
Comment (FabianKlimpel) [inline Fatras/include/ActsFatras/EventData/Particle.hpp:143]:
```suggestion
    mom4[Acts::eMom1] = m_absMomentum * m_unitDirection[Acts::ePos1];
```

File: Fatras/include/ActsFatras/EventData/Particle.hpp [hunk 5]
```diff
@@ -132,30 +130,26 @@ class Particle {
...
+    mom4[Acts::eMom2] = m_absMomentum * m_unitDirection[2];
```
Comment (FabianKlimpel) [inline Fatras/include/ActsFatras/EventData/Particle.hpp:144]:
```suggestion
    mom4[Acts::eMom2] = m_absMomentum * m_unitDirection[Acts::ePos2];
```

File: Tests/UnitTests/Core/EventData/ParameterSetTests.cpp [hunk 6]
```diff
@@ -415,7 +415,7 @@ BOOST_AUTO_TEST_CASE(parset_copy_assignment_tests) {
   double loc1 = -0.2;
   double phi = 0.3 * M_PI;  // this should be within [-M_PI,M_PI) to avoid
                             // failed tests due to angle range corrections
-  ActsVectorD<3> first_parValues(loc0, loc1, phi);
+  Vector3D first_parValues(loc0, loc1, phi);
```
Comment (FabianKlimpel) [inline Tests/UnitTests/Core/EventData/ParameterSetTests.cpp:418]:
```suggestion
  Vector3D firstParValues(loc0, loc1, phi);
```

File: Tests/UnitTests/Core/EventData/ParameterSetTests.cpp [hunk 7]
```diff
@@ -858,7 +858,7 @@ BOOST_AUTO_TEST_CASE(free_parset_copy_assignment_tests) {
   double x = 0.5;
   double y = -0.2;
   double z = 0.3;
-  ActsVectorD<3> first_parValues(x, y, z);
+  Vector3D first_parValues(x, y, z);
```
Comment (FabianKlimpel) [inline Tests/UnitTests/Core/EventData/ParameterSetTests.cpp:861]:
```suggestion
  Vector3D firstParValues(x, y, z);
```

**Comment by msmk0** [comment]
Thanks for catching that. Should be all in now.

**Comment by msmk0** [comment]
@FabianKlimpel I solved the merge conflicts. Could you re-approve?

**Comment by FabianKlimpel** [comment]
> @FabianKlimpel I solved the merge conflicts. Could you re-approve?

Here it says that is still approved.

**Comment by msmk0** [comment]
Even better. Then I just wait for the CI to finish. Thanks.

**Comment by FabianKlimpel** [comment]
> Even better. Then I just wait for the CI to finish. Thanks.

I think I saw this before but never really tested whether this really counts as approve, so please test it and hope that it doesn't work. Otherwise this would allow dangerous things.
